### PR TITLE
Switch return type of main from 'void' to 'int'

### LINKED ISSUE #56504: Switch return type of main to 'int'
## Introduction

The application entry point used by zephyr, `main`, doesn't currently have any useful return value. As a result, it has been declared to return `void`. I propose to change that to `int` as required by the C and C++ standards for "hosted" applications.

### Problem description

Zephyr avoids compiler issues when using a `void` return type of `main` by a combination of three methods:

 1. Compiling with `-ffreestanding`
 2. Special casing C++ applications and having those use a return type of `int`.
 3. Adding compiler flags, `-Wno-main`  and `-Wno-main-return-type`

Switching to a return type of  `int` removes the need for using any of these.

Removing the `-ffreestanding` compiler parameter, as proposed in #54638, will enable significantly better error detection in code using standard C library functions as well as improve generated code quality as the compiler will be able to use knowledge about C library semantics during compilation. However, clang emits an error (even with `-Wno-main-return-type`) when building without the `-ffreestanding` flag. 

### Proposed change

Applications will need to switch the return type of their `main` entry point from `void` to `int`. Zephyr will continue to ignore the return value meaning that there will be no change in how the system operates.

## Detailed RFC

PR #54628 shows the sequence of tasks needed to accomplish this goal:

 1. Updating documentation describing the declaration of  `main`
 2. Declaring main in kernel/init.c to return `int` and then explicitly ignoring the value
 3. Switching the definition of main in all in-tree samples and tests to return `int`.

### Proposed change (Detailed)

Most of the changes for function definitions are performed with a coccinelle script, with a handful of manual fixups.

### Dependencies

This should have limited effect on other parts of the system, except for pending additions of new tests and samples. As no behavior change is proposed, this should not affect test results anywhere.

### Concerns and Unresolved Questions

Should the return value of  `main` actually do something? If so, what should that be? The current PR uniformly returns `0`, so we could, in the future, define non-zero return values to do something different.

## Alternatives

The obvious alternative is to do nothing. However, leaving the return value as `void` means that we can never remove the `-ffreestanding` compiler flag as clang will generate an error on a `void` valued `main` definition.

### PR DESCRIPTION
**Pull request by keith-packard**
As both C and C++ standards require applications running under an OS to
return 'int', adapt that for Zephyr to align with those standard. This also
eliminates errors when building with clang when not using -ffreestanding,
and reduces the need for compiler flags to silence warnings for both clang
and gcc.

---

Related discussion: https://github.com/zephyrproject-rtos/zephyr/issues/54336#issuecomment-1421528758

### CODE REVIEW

**Comment by stephanosio** [comment]
FYI @andyross @peter-mitsis @teburd @nashif @cfriedt

**Review by cfriedt — DISMISSED** [review]
Just to clarify, I think this should go in after the release, but 100% agree. Thanks!

**Comment by keith-packard** [comment]
> It makes sense, arguably a breaking API change though and needs to go through that? What happens with all the existing applications using void main(void)?

Yeah, it's "kinda" an API change. `main` isn't declared in any Zephyr header. If we don't adjust the compiler flags, then there's actually no impact on applications at all. The goal is to allow compiling using clang without `-ffreestanding`, and that requires that main return int.

**Comment by keith-packard** [comment]
> Just to clarify, I think this should go in after the release, but 100% agree. Thanks!

Glad to see @stephanosio adjusted the milestone to 3.4.0 -- I wasn't sure what label to apply to make sure this wasn't considered for this upcoming release.

**Review by de-nordic — DISMISSED** [review]
OK for smp_svr sample.

**Review by nordicjm — DISMISSED** [review]
Will need a release note entry added and needs to be brought up at the architecture working group but change itself is good

File: samples/bluetooth/beacon/src/main.c
Comment (Thalley) [inline samples/bluetooth/beacon/src/main.c:88]: One comment that may only apply for BT, but for the BT subsystem we usually keep an empty line after an ending `}`, so this would be

```suggestion
	}
	
	return 0;
```

File: samples/bluetooth/broadcast_audio_sink/src/main.c
Comment (Thalley) [inline samples/bluetooth/broadcast_audio_sink/src/main.c:213]: Wouldn't it make more sense to return the error? 
```suggestion
		return err;
```

This applies many places

File: samples/subsys/mgmt/updatehub/src/main.c
Comment (nandojve) [inline samples/subsys/mgmt/updatehub/src/main.c:106]: Hi @keith-packard ,

I noted that these two init errors not return a value. Should those return a -errno code ?

**Review by nandojve — CHANGES_REQUESTED** [review]

**Comment by peter-mitsis** [comment]
Aside from the occasional questionable return value (which others have already pointed out), I do not have any fundamental or design issues with what has been proposed.

File: samples/bluetooth/beacon/src/main.c
Comment (keith-packard) [reply samples/bluetooth/beacon/src/main.c:88]: These changes were done using coccinelle; I see a bunch of value in not messing with the results of that where possible as the tool is less likely to generate mistakes than manual editing? Perhaps a manual clean up pass in the series which adjusts formatting could be done though?

File: samples/bluetooth/broadcast_audio_sink/src/main.c
Comment (keith-packard) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: It might, but Zephyr explicitly ignores the return value, so it really doesn't matter in practice.

File: samples/subsys/mgmt/updatehub/src/main.c
Comment (keith-packard) [reply samples/subsys/mgmt/updatehub/src/main.c:106]: Zephyr explicitly ignores the return value from main(). Maybe the documentation patch should make it explicit that a return value of zero is expected?

Comment (nandojve) [reply samples/subsys/mgmt/updatehub/src/main.c:106]: Since Zephyr don't care about the return value you can ignore my previous comment. Documentation may be good to avoid same questions on the communication channels.

**Review by nandojve — DISMISSED** [review]

File: samples/subsys/mgmt/updatehub/src/main.c
Comment (keith-packard) [reply samples/subsys/mgmt/updatehub/src/main.c:106]: good plan. I've added a sentence to both the c and c++ documents indicating that Zephyr ignores the return value from main and that applications should return zero.

File: samples/bluetooth/beacon/src/main.c
Comment (Thalley) [reply samples/bluetooth/beacon/src/main.c:88]: > These changes were done using coccinelle; 

It does not have an option to specify empty line after ending bracets?

File: samples/bluetooth/broadcast_audio_sink/src/main.c
Comment (Thalley) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: If we ignore the return value, then why even bother adding the return type `int`? 

If we run these samples as Linux applications or part of CI, we definitely don't want to ignore the return value of the `main` function, as we want to catch errors. 

IMO blindly returning `0` in all cases is incorrect.

Comment (keith-packard) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: Totally aligned with your remarks; it seems like an arbitrary change that has zero benefit and the potential for confusion. I've tried to clarify some of this in the doc updates; perhaps some additional wording would be helpful?

The reason for the switch is because the C and C++ standards *require* that `main` be defined as either  `int main(void)` or `int main(int argc, char **argv)` for "hosted" environments. Right now, the GCC and Clang compiler configurations include special compiler flags (`-Wno-main` and `-Wno-main-return-type`) to suppress the compiler warnings. In the future, when the -ffreestanding flag gets removed, the Clang warning turns into a Clang fatal error.

Feel free to complain to the ISO C and C++ committees about their standards, or to the Clang and GCC compiler writers for enforcing them.

Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

Comment (Thalley) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: @keith-packard I guess it will be too much effort for you to change the `return 0` to `return err` where applicable? 

> Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

As mentioned above, it does have some meaning already in certain cases

Comment (stephanosio) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: > If we ignore the return value, then why even bother adding the return type int?

Standard conformance -- some compilers refuse to compile `void main(void)` because that is not one of the allowed types for the `main` function defined by the C standard.

> Alternatively, maybe we should change the Zephyr OS so that the return value from main has some meaning?

The return value could be used by the Zephyr kernel to do something useful (e.g. reboot if the main function returns a non-zero value with the assumption that the application initialisation failed); but, that is beyond the scope of this PR and can be addressed separately if we deem something like that to be necessary in the future.

> If we run these samples as Linux applications or part of CI, we definitely don't want to ignore the return value of the `main` function, as we want to catch errors.

But can we currently run these samples as Linux applications? If not, I do not see how that argument is valid at this time -- the return values can be updated accordingly when the sample is updated to be able to run as a Linux application, or when the Zephyr kernel is updated to actually make use of the return value -- what Zephyr kernel will do with the return value is still an open question and beyond the scope of this PR as described above.

Comment (Thalley) [reply samples/bluetooth/broadcast_audio_sink/src/main.c:213]: > But can we currently run these samples as Linux applications?

Yes :) In Nordic we do in fact also run the samples as part of CI, but given that we haven't "needed" the return values yet (we just parse the stdout data), it's a fair argument to not add those in this PR. 

If we don't add the (proper) return values in this PR, I'll just add them later for the Bluetooth samples myself :)

File: doc/develop/languages/c/index.rst
Comment (gregshue) [inline doc/develop/languages/c/index.rst:18]: Please describe the API policy for `main()` rather than describe current Zephyr implementation ("Zephyr ignores the return value of main").  Something like the statement below would work well.  Ideally this would be documented as a public API (even if the actual declaration was conditionally-compiled out):

- All non-zero return values from `main()` are reserved.
- All meanings of return values from `main()` are reserved.

**Comment by gregshue** [comment]
Regarding return value usage and meaning:

Bootloaders are a ubiquitous example of an application needing to cleanly shut down the execution environment before handing the HW and CPU(s) off to another executable. To maintain a composable, scalable, and extensible architecture this clean shut down maps to a reverse `SYS_INIT()` which should be executed by the same thread that called `SYS_INIT()`. Thus, returning from `main()` can be very useful to trigger for a shutdown.

AFAICT, the shutdown sequence either succeeds or should cause a catastrophic failure (i.e., `k_panic()`/`z_except_reason()`). IF true, then this sequence has no need for the return value - returning from main() is sufficient to trigger the shutdown sequence. Shutdown parameters (e.g., the pointer to jump to upon shutdown, whether or not to reset upon shutdown, etc.) may be specific to the subsystems involved so shouldn't expect to be encodable in an `int` or passed via the return value.

It may be wise to leave meaning and usage of non-zero return values for `native_posix`, `unit_testing`, and/or emulation environments rather than let the Zephyr kernel try to take advantage of it.

File: doc/develop/languages/c/index.rst
Comment (keith-packard) [reply doc/develop/languages/c/index.rst:18]: Yeah, you're right, the docs should state the policy not just the current implementation. Once that policy has been agreed upon, we can update the docs to match as part of this PR.

**Comment by galak** [comment]
If we merge this, we need to "revert" (and similar changes in other files for `cmake/compiler/arcmwdt/compiler_flags.cmake`):

commit 0f692f5b84db8ef61f1e85c6ef25278966dc3329
Author: Keith Packard <keithp@keithp.com>
Date:   Mon Feb 6 21:10:29 2023 -0800

    compiler/clang: ignore main-return-type warning for clang

**Comment by mbolivar-nordic** [comment]
Arch WG:

- This is a treewide change, so it requires an RFC issue and other processes to be followed (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes), but no objections were raised
- https://github.com/zephyrproject-rtos/zephyr/pull/54525 should be reverted if this is merged
- prior discussion in https://github.com/zephyrproject-rtos/zephyr/pull/21100#pullrequestreview-325259287 
- discussion around to start, requiring 0 return value and reserve other values for future use (https://github.com/zephyrproject-rtos/zephyr/pull/54628#issuecomment-1431921596)
- request to verify consequences of not changing the return type and document migration steps in release notes

**Comment by keith-packard** [comment]
> Arch WG:
> 
>     * This is a treewide change, so it requires an RFC issue and other processes to be followed (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#treewide-changes), but no objections were raised

#56504 

>     * [compiler/clang: ignore main-return-type warning for clang #54525](https://github.com/zephyrproject-rtos/zephyr/pull/54525) should be reverted if this is merged

These flags are removed as a part of this PR.

>     * request to verify consequences of not changing the return type and document migration steps in release notes

The only consequence of not changing that I'm aware of is that we will not be able to remove the `-ffreestanding` compiler flag when using clang as that has an un-skippable error for a non-int main return type (even if `-Wno-main-return-type` is specified). Consequences of the `-ffreestanding` compiler flag are discussed in PR #54638

**Comment by keith-packard** [comment]
(I've rebased and fixed the conflicts today)

**Comment by keith-packard** [comment]
> If we merge this, we need to "revert" (and similar changes in other files for `cmake/compiler/arcmwdt/compiler_flags.cmake`):

This change is already part of this PR.

File: doc/develop/languages/c/index.rst
Comment (keith-packard) [reply doc/develop/languages/c/index.rst]: I've changed the wording to require a return value of zero from main (at least for now).

**Review by jfischer-no — DISMISSED** [review]

**Comment by carlescufi** [comment]
Setting DNM since this requires agreement before merging.

**Review by carlescufi — DISMISSED** [review]
Please add an entry in the major changes sections of the release notes in:
https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.4.rst

File: samples/drivers/display/src/main.c
Comment (ifyall) [inline samples/drivers/display/src/main.c:1]: Per the documentation we are adding in https://github.com/zephyrproject-rtos/zephyr/pull/54628/commits/1b765d022204ae6fa99835a19ffd54facc5fbe5f, a non-zero return type is not allowed at this time. This file doesn't adhere to that. I won't comment on the posix return case, as I am not familiar enough with the standard. However, should the non-posix flow return 0 in compliance with the doc? If not, then we should have comments in this file indicating why we are allowing the exception.

**Review by ifyall — CHANGES_REQUESTED** [review]

File: samples/drivers/display/src/main.c
Comment (keith-packard) [reply samples/drivers/display/src/main.c:1]: oops! All fixed.

**Comment by keith-packard** [comment]
> Please add an entry in the major changes sections of the release notes in: https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/releases/release-notes-3.4.rst

Done.

**Review by jfischer-no — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]
LGTM

**Comment by stephanosio** [comment]
Needs a rebase.

**Comment by keith-packard** [comment]
> Needs a rebase.

Yup, all done. No significant conflicts found.

**Review by ifyall — DISMISSED** [review]

**Review by microbuilder — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Comment by stephanosio** [comment]
Will merge once approved in the TSC meeting today.

**Review by jhedberg — DISMISSED** [review]

**Review by nordicjm — COMMENTED** [review]
@aescolar seems like babblesim will need updating:
```
/*
 * This function will be called (in SW context) as the Zephyr application main
 */
typedef void (*bst_test_main_t)(void);
```
And various uses in tests:
```
./tests/bsim/bluetooth/host/privacy/peripheral/src/main.c:void main(void)
./tests/bsim/bluetooth/host/privacy/device/src/test_undirected_main.c:void main(void)
./tests/bsim/bluetooth/host/privacy/central/src/main.c:void main(void)
./tests/bsim/bluetooth/host/gatt/settings/src/main.c:void main(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_conn_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_conn_privacy_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_syncer.c:static void main_per_adv_long_data_syncer(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_conn_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_conn_privacy_advertiser(void)
./tests/bsim/bluetooth/host/adv/periodic/src/per_adv_advertiser.c:static void main_per_adv_long_data_advertiser(void)
./tests/bsim/bluetooth/host/adv/encrypted/css_sample_data/src/main.c:void main(void)
```

**Review by nordicjm — CHANGES_REQUESTED** [review]
Following seem to have been missed in conversion:
```
./tests/drivers/build_all/mfd/src/main.c
./samples/subsys/logging/ble_backend/src/main.c
./samples/sensor/proximity_polling/src/main.c
./samples/fuel_gauge/max17048/src/main.c
./samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c
./samples/drivers/smbus/src/main.c
./samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c
./samples/bluetooth/bthome_sensor_template/src/main.c
```
Documentation in ./arch/posix/include/posix_cheats.h should also be updated around this part:
```
#if defined(__cplusplus) && defined(CONFIG_CPP_MAIN)
/* To be able to define main() in C++ code we need to have its prototype
 * defined somewhere visibly. Otherwise name mangling will prevent the linker
 * from finding it. Zephyr assumes a void main(void) prototype and therefore
 * this will be the prototype after renaming:
 */
extern "C" int _posix_zephyr_main(void);
#endif
```

**Comment by stephanosio** [comment]
Removing DNM since the proposal in #56504 has been approved by the TSC.

**Review by cfriedt — DISMISSED** [review]

**Review by nordicjm — CHANGES_REQUESTED** [review]
Following are missing changes:
```
./tests/drivers/build_all/mfd/src/main.c:void main(void)
./samples/subsys/logging/ble_backend/src/main.c:void main(void)
./samples/sensor/proximity_polling/src/main.c:void main(void)
./samples/hello_world/src/main.c:void main(void)
./samples/fuel_gauge/max17048/src/main.c:void main(void)
./samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c:void main(void)
./samples/drivers/smbus/src/main.c:void main(void)
./samples/boards/stm32/power_mgmt/standby_shutdown/src/main.c:void main(void)
./samples/bluetooth/bthome_sensor_template/src/main.c:void main(void)
```

**Comment by carlescufi** [comment]
@keith-packard there are no TSC objections to merging this. Could you rebase and address the missing ones that @nordicjm found? Thanks!

**Comment by keith-packard** [comment]
> @keith-packard there are no TSC objections to merging this. Could you rebase and address the missing ones that @nordicjm found? Thanks!

rebased, repaired and re-pushed.

**Review by nordicjm — APPROVED** [review]
This will need to be removed but it can be done in a later commit, no point holding this up anymore:
```
config CPP_MAIN
        bool "C++ main() function definition"
        help
          This option instructs the Zephyr kernel to call the 'int main(void)'
          instead of the 'void main(void)', which is the default main() type
          for Zephyr.

          C++ does not allow the main() to be defined with 'void' return type,
          and any applications defining its main() in a C++ source file must
          enable this option.
```
CC @stephanosio 
And ./arch/posix/include/posix_cheats.h too

**Review by stephanosio — APPROVED** [review]

**Review by jfischer-no — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Comment by keith-packard** [comment]
I just realized that we can add `-Wmain` to the compiler flags if we want to catch applications using the wrong type for main. This works with gcc even with `-ffreestanding`, but doesn't appear to work with clang with that flag.

**Pull request merged**
